### PR TITLE
Add schema generation and verification

### LINKED ISSUE #1090: Schema generation/validation
Related to https://github.com/paritytech/ink/issues/1695
Introduce new commands for schema generation and validation:
- `cargo contract schema` - generate a schema file
- `cargo contract schema --validate/-v <path>` - validate local metadata file against schema specified

### PR DESCRIPTION
**Pull request by SkymanOne**
## Summary
Closes #1090 and https://github.com/paritytech/ink/issues/1564
- [n] y/n | Does it introduce breaking changes?
- [y] y/n | Is it dependent on the specific version of `ink` or `pallet-contracts`? - 5.0.0-alpha

Introudeces schema generation and verfication commands.

## Description

This PR automated proces for metadata schema generation.

Whenever there are changes to the schema, if `cargo-contract` targets the latest version of `ink_metadata` the scheme can be generated with `cargo contract schema-generate inner/outer`. This will print json to STDOUT.

And here is the sample of how to verify the metadata against specified schema file:

Bundle:

```
cargo contract verify-schema --schema /Users/skymanone/Projects/ink/crates/metadata/contract_schema.json --bundle target/ink/flipper.contract

Successfully verified metadata in `target/ink/flipper.contract` against schema `/Users/skymanone/Projects/ink/crates/metadata/contract_schema.json`!
```

Metadata file:
```
cargo contract verify-schema --schema /Users/skymanone/Projects/ink/crates/metadata/contract_schema.json --metadata target/ink/flipper.json

Successfully verified metadata in `target/ink/flipper.json` against schema `/Users/skymanone/Projects/ink/crates/metadata/contract_schema.json`!
```

## Checklist before requesting a review
- [x] My code follows the style guidelines of this project
- [x] I have added an entry to `CHANGELOG.md`
- [x] I have commented my code, particularly in hard-to-understand aras
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: crates/cargo-contract/src/cmd/schema.rs [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+use schemars::schema_for;
+
+#[derive(Debug, Clone, Default, clap::ValueEnum)]
+#[clap(name = "metadata")]
+enum Metadata {
+    /// Represents the outer schema format of the contract
+    #[clap(name = "outer")]
+    #[default]
+    /// Represents the inner schema format of the contract.
+    /// Contains specification of the ink! contract.
```
Comment (smiasojed) [inline crates/cargo-contract/src/cmd/schema.rs:26]: this comment should be moved, from outer to inner

File: crates/cargo-contract/src/cmd/schema.rs [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+                    "Failed to deserialize contract bundle {}",
+                    path.display()
+                ))?;
+            contract_metadata.remove_source_wasm_attribute();
+
+            metadata = serde_json::value::to_value(contract_metadata)?;
+            metadata_source = path.display().to_string();
+        }
+
+        // 1a. Read metadata file
```
Comment (smiasojed) [inline crates/cargo-contract/src/cmd/schema.rs:103]:
```suggestion
        // 1b. Read metadata file
```

File: crates/cargo-contract/src/cmd/schema.rs [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+                String::from("Error during schema validation:\n"),
+                |acc, e| format!("{}\n{}", acc, e),
+            );
+            anyhow!(error_msg)
+        })?;
+
+        Ok(SchemaVerificationResult {
+            is_verified: true,
+            metadata_source,
+            schema: self.schema.display().to_string(),
```
Comment (smiasojed) [inline crates/cargo-contract/src/cmd/schema.rs:146]: Why are we keeping the schema as a string in the structure? Perhaps we should convert it to a string during struct serialization?

Comment (SkymanOne) [reply crates/cargo-contract/src/cmd/schema.rs:146]: This is a path to a schema file

File: crates/cargo-contract/src/main.rs [hunk 4]
```diff
@@ -142,6 +144,12 @@ enum Command {
     /// workspace.
     #[clap(name = "verify")]
     Verify(VerifyCommand),
+    /// Generates schema from the current metadata specification.
+    #[clap(name = "generate-schema")]
```
[the remark this answers is not in the document]
Comment (SkymanOne) [reply crates/cargo-contract/src/main.rs:148]: Too many subcommands IMO: `cargo contract schema ...`

File: crates/cargo-contract/src/cmd/schema.rs [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+    VerbosityFlags,
+};
+use jsonschema::JSONSchema;
+use schemars::schema_for;
+
+#[derive(Debug, Clone, Default, clap::ValueEnum)]
+#[clap(name = "metadata")]
+enum Metadata {
+    /// Represents the outer schema format of the contract
+    #[clap(name = "outer")]
```
[the remark this answers is not in the document]
Comment (SkymanOne) [reply crates/cargo-contract/src/cmd/schema.rs:23]: These are the terms that were originally used

[the remark this answers is not in the document]
Comment (smiasojed) [reply crates/cargo-contract/src/cmd/schema.rs:23]: If these terms are commonly used by developers in the ecosystem, I'm fine with it. However, if they are our internal terms, I would suggest changing them.

File: crates/cargo-contract/src/main.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (smiasojed) [reply crates/cargo-contract/src/main.rs:148]: I think `info` and `call` have more arguments, but it's up to you whether to leave them as they are or make changes.

File: crates/cargo-contract/src/cmd/schema.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (SkymanOne) [reply crates/cargo-contract/src/cmd/schema.rs:23]: See https://github.com/paritytech/ink/tree/master/crates/metadata

**Review by smiasojed — APPROVED** [review]
LGTM

**Pull request merged**